### ISSUE TITLE
completeMultipartUpload completes the upload but reports NoSuchKey error

### ISSUE THREAD

**Issue by kyleburnett** [minio/minio#4727]
## Expected Behavior
Theoretically, the test.js file in the gist provided should run forever.

## Current Behavior
However, eventually it outputs the following error:

```
NoSuchKey: The specified key does not exist.
    at Request.extractError (/Users/kyle/Documents/scripts/minio-test/node_modules/aws-sdk/lib/services/s3.js:568:35)
    at Request.callListeners (/Users/kyle/Documents/scripts/minio-test/node_modules/aws-sdk/lib/sequential_executor.js:105:20)
    at Request.emit (/Users/kyle/Documents/scripts/minio-test/node_modules/aws-sdk/lib/sequential_executor.js:77:10)
    at Request.emit (/Users/kyle/Documents/scripts/minio-test/node_modules/aws-sdk/lib/request.js:682:14)
    at Request.transition (/Users/kyle/Documents/scripts/minio-test/node_modules/aws-sdk/lib/request.js:22:10)
    at AcceptorStateMachine.runTo (/Users/kyle/Documents/scripts/minio-test/node_modules/aws-sdk/lib/state_machine.js:14:12)
    at /Users/kyle/Documents/scripts/minio-test/node_modules/aws-sdk/lib/state_machine.js:26:10
    at Request.<anonymous> (/Users/kyle/Documents/scripts/minio-test/node_modules/aws-sdk/lib/request.js:38:9)
    at Request.<anonymous> (/Users/kyle/Documents/scripts/minio-test/node_modules/aws-sdk/lib/request.js:684:12)
    at Request.callListeners (/Users/kyle/Documents/scripts/minio-test/node_modules/aws-sdk/lib/sequential_executor.js:115:18)
    at Request.emit (/Users/kyle/Documents/scripts/minio-test/node_modules/aws-sdk/lib/sequential_executor.js:77:10)
    at Request.emit (/Users/kyle/Documents/scripts/minio-test/node_modules/aws-sdk/lib/request.js:682:14)
    at Request.transition (/Users/kyle/Documents/scripts/minio-test/node_modules/aws-sdk/lib/request.js:22:10)
    at AcceptorStateMachine.runTo (/Users/kyle/Documents/scripts/minio-test/node_modules/aws-sdk/lib/state_machine.js:14:12)
    at /Users/kyle/Documents/scripts/minio-test/node_modules/aws-sdk/lib/state_machine.js:26:10
    at Request.<anonymous> (/Users/kyle/Documents/scripts/minio-test/node_modules/aws-sdk/lib/request.js:38:9)
{ [NoSuchKey: The specified key does not exist.]
  cause:
   { [NoSuchKey: The specified key does not exist.]
     message: 'The specified key does not exist.',
     code: 'NoSuchKey',
     region: null,
     time: Wed Jul 26 2017 14:43:11 GMT-0400 (EDT),
     requestId: '14D4F593D38DBDEA',
     extendedRequestId: undefined,
     cfId: undefined,
     statusCode: 404,
     retryable: false,
     retryDelay: 32.2675014147535 },
  isOperational: true,
  code: 'NoSuchKey',
  region: null,
  time: Wed Jul 26 2017 14:43:11 GMT-0400 (EDT),
  requestId: '14D4F593D38DBDEA',
  extendedRequestId: undefined,
  cfId: undefined,
  statusCode: 404,
  retryable: false,
  retryDelay: 32.2675014147535 }
Object exists!
```

The completeMultipartUpload call appears to fail, but when I try to "head" the object, it says it exists, which leads me to believe that completeMultipartUpload was successful after all.

## Steps to Reproduce (for bugs)
Here is a gist: https://gist.github.com/kyleburnett/bb85c386162e2a32c6d092f7898ea6a4
1. Run `npm install` to install dependencies.
2. Run `node setup.js` to generate random data.
3. Replace "your accessKeyId here", "your secretAccessKey here", and "your endpoint here" in `test.js`
4. Run `test.js`.

## Context
I'm trying to upload many files concurrently. I originally discovered this issue because unit tests were failing every 20 or so runs.

## Your Environment
* On a Vagrant VM:
  * Minio: 2017-06-13T19:01:01Z
  * Docker: v1.11.1
  * OS: Ubuntu 14.04.4
* Host
  * Node.js: v4.4.7
  * OS: macOS Sierra v10.12.5

**Comment by harshavardhana** [comment]
> On a Vagrant VM:
> Minio: 2017-06-13T19:01:01Z
> Docker: v1.11.1
> OS: Ubuntu 14.04.4

Can you use the latest release @kyleburnett and let us know?

**Comment by kyleburnett** [comment]
I just used the latest version. Here is the `minio version` info:
```
/ # minio version
Version: 2017-07-24T18:27:35Z
Release-Tag: RELEASE.2017-07-24T18-27-35Z
Commit-ID: 461d60095df56cdb1370f20135429a5a24b400b3
```

Same result.

**Comment by brendanashworth** [comment]
@kyleburnett I think the issue lies in your test case. I can confirm the error happens on my machine, but for me it's gone in [this cleaned up test case](https://gist.github.com/brendanashworth/dc35a9286f4bf58c65840a4a029ceb4f). Does this work for you? I just cleaned up the promise code and removed the `this` bindings.

**Comment by kyleburnett** [comment]
I updated [my test case](https://gist.github.com/kyleburnett/bb85c386162e2a32c6d092f7898ea6a4) to look more like yours with one difference in particular that I hope will help illustrate why I'm seeing this issue. The parts that are prefixed with "Setup:" are just setting things up. These are the calls to create the multipart upload and upload the data in a single part. For my case, these calls are being done in their own separate API routes. Then a third API route is handling what I've put into the `listAndComplete` function (i.e., many concurrent uploads completing at the same time). Is there any reason why I shouldn't be able to use `listParts` and `completeMultipartUpload` together?

I added the `headObject` part back because I think it shows that the object exists even though the error thrown by `completeMultipartUpload` says otherwise.

Another way to think about this is I have a bunch of clients concurrently uploading parts independently of one another and after each part is uploaded, I need to check whether I have all of the parts and if I do, I complete the multipart upload. In your code, you separated out the `listParts` and `completeMultipartUpload`, which does indeed seem to fix the issue, however, the reason they are together in my example is because the part uploads are actually happening concurrently, and thus, multiple calls to `listParts` and `completeMultipartUpload` can be happening at the same time (i.e., multiple calls to `listAndComplete`). I’m guessing that it is in fact this concurrency that is causing the issue in minio, however, I can’t separate the two because of how the uploads are actually being sent to the server.

**Comment by brendanashworth** [comment]
Thank you @kyleburnett for the clarification. I can confirm this is a bug. Under certain conditions, a `CompleteMultipartUpload` request will return an error after failing through [this code branch](https://github.com/minio/minio/blob/ec5293ce29ef03726c7174d9e3304bc156259c4e/cmd/fs-v1-multipart.go#L897-L899). It is [not covered](https://codecov.io/gh/minio/minio/src/ec5293ce29ef03726c7174d9e3304bc156259c4e/cmd/fs-v1-multipart.go#L898) in our code coverage. I am unsure about what exact conditions bring about this situation, although the test case above replicates the bug. I do not think it's an issue with `aws-sdk-js`.

**Comment by kyleburnett** [comment]
Thanks @brendanashworth for investigating this. For now, would a safe workaround be to perform a headObject call to check that the object exists? In other words, if `completeMultipartUpload` throws a 'NoSuchKey' error, I would call headObject. If the object exists, I ignore the error and move on.

Something like this:
```
function listAndComplete(obj) {
    return s3.listPartsAsync(...).then(function() {
        ...

        return s3.completeMultipartUploadAsync(...).catch(function(e_orig) {
            if (e_orig.code === 'NoSuchKey') {
                return s3.headObjectAsync(...).then(function() {
                    // ignore error
                }).catch(function(e) {
                    throw e_orig;
                });
            } else {
                throw e_orig;
            }
        });
    });
}
```

**Comment by brendanashworth** [comment]
@kyleburnett that should be ok as long as you're not editing the same object in between, i.e. triggering a race condition. I'll try and see that this is fixed soon though.

**Comment by brendanashworth** [comment]
I've found the fix, and it will find its way into the code base soon. Here's a paste of the cause of the issue, for what its worth:
```
    The race condition occurs when multiple deleteFile()s are called on the
    same parent directory, but different child files. Because deleteFile()
    recursively removes parent directories if they are empty, but
    deleteFile() errors if the selected deletePath does not exist, there was
    an opportunity for a race condition. The two processes would remove the
    child directories successfully, then depend on the parent directory
    still existing. In some cases this is an invalid assumption, because
    other processes can remove the parent directory beforehand. This commit
    changes deleteFile() to not upstream an error if one occurs, because the
    only required error should be from the immediate deletePath, not from a
    parent path.
    
    In the specific bug report, multiple CompleteMultipartUpload requests
    would launch multiple deleteFile() requests. Because they chain up on
    parent directories, ultimately at the end, there would be multiple
    remove files for the ultimate parent directory,
    .minio.sys/multipart/{bucket}. Because only one will succeed and one
    will fail, an error would be upstreamed saying that the file does not
    exist, and the CompleteMultipartUpload code interpreted this as
    NoSuchKey, or that the object/part id doesn't exist. This was faulty
    behavior and is now fixed.
```

**Comment by kyleburnett** [comment]
Thanks, @brendanashworth, I'm looking forward to the fix.

This is a awesome project - keep up the great work!

**Comment by brendanashworth** [comment]
@kyleburnett the fix has landed in https://github.com/minio/minio/commit/aeafe668d8b6d25caac671d59e2b0f0473ce35d0 and will be in the next release! Thank you for reporting this.

**Comment by harshavardhana** [comment]
> @kyleburnett the fix has landed in aeafe66 and will be in the next release! Thank you for reporting this.

A new release has been made with the changes https://dl.minio.io/server/minio/release/linux-amd64/
